storage/inspector/src/main.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use anyhow::{ensure, Result};
+use libra_logger::prelude::*;
+use libradb::LibraDB;
+use std::path::PathBuf;
+use transaction_builder::get_transaction_name;
+
+use libra_crypto::hash::{CryptoHash, EventAccumulatorHasher};
+use libra_crypto::HashValue;
+use libra_types::contract_event::ContractEvent;
+use libra_types::proof::accumulator::InMemoryAccumulator;
+use structopt::StructOpt;
+
+#[derive(Debug, StructOpt)]
+struct Opt {
+    #[structopt(long, parse(from_os_str))]
+    db: PathBuf,
+}
+
+/// Print out latest information stored in the DB.
+fn print_head(db: &LibraDB) {
+    let version = db
+        .get_latest_version()
+        .expect("get_latest_version throw error");
+
+    info!("Latest Version: {}", version);
+
+    let si = db
+        .get_startup_info()
+        .expect("Can't get startup info")
+        .expect("StartupInfo is empty, database is empty.");
+
+    info!(
+        "The latest ledger info: {}",
+        si.latest_ledger_info.ledger_info()
+    );
+
+    info!("Signatures: {:?}", si.latest_ledger_info.signatures());
+
+    info!(
+        "Current Validator Set: {}",
+        si.latest_validator_set
+            .expect("Unable to determine validator set, DB incorrect")
+    );
+
+    let tx = db
+        .get_transaction_with_proof(version, version, false)
+        .expect("Unable to load latest TXN");
+    info!(
+        "Current Transaction: {}",
+        tx.transaction.format_for_client(get_transaction_name)
+    );
+}
+
+/// List all TXNs from DB and verify their proofs
+fn verify_txns(db: &LibraDB) -> Result<()> {
+    let si = db
+        .get_startup_info()
+        .expect("Can't get startup info")
+        .expect("StartupInfo is empty, database is empty.");
+    let ledger_info = si.latest_ledger_info.ledger_info();
+
+    let latest_version = ledger_info.version();
+
+    for n in (1..latest_version).rev() {
+        let tx = db.get_transaction_with_proof(n, latest_version, true)?;
```

Comment:
Again it's a waste of CPU and IO to do this for each transaction. Should recursively verify down the tree instead. But we can probably do it seperately.

Reply:
also we need to note here by getting all events on a txn here, you are not verifying the correctness of the event accumulator, because it doesn't return the per event proofs. We need to again recursively go down each event accumulator if you are seriously verifying everything is in place.